client/engine/messageservice/p2p-message-service/service.go

```diff
@@ -91,26 +83,100 @@ func NewMessageService(ip string, port int, me types.Address, pk []byte) *P2PMes
-		// Create a buffer stream for non blocking read and write.
-		raw, err := reader.ReadString(DELIMITER)
+	ms.p2pHost.SetStreamHandler(PROTOCOL_ID, ms.msgStreamHandler)
+	ms.p2pHost.SetStreamHandler(PEER_EXCHANGE_PROTOCOL_ID, ms.peerExchangeHandler)
 
-		// An EOF means the stream has been closed by the other side.
-		if errors.Is(err, io.EOF) {
-			stream.Close()
-			return
-		}
-		ms.checkError(err)
-		m, err := protocols.DeserializeMessage(raw)
-		ms.checkError(err)
-		ms.toEngine <- m
+	return ms
+}
+
+// HandlePeerFound is called by the mDNS service when a peer is found.
+func (ms *P2PMessageService) HandlePeerFound(pi peer.AddrInfo) {
+	ms.p2pHost.Peerstore().AddAddr(pi.ID, pi.Addrs[0], peerstore.PermanentAddrTTL)
+	stream, err := ms.p2pHost.NewStream(context.Background(), pi.ID, PEER_EXCHANGE_PROTOCOL_ID)
+	ms.checkError(err)
+	ms.sendPeerInfo(stream)
+	ms.receivePeerInfo(stream)
+	stream.Close()
+}
+
+func (ms *P2PMessageService) msgStreamHandler(stream network.Stream) {
+	reader := bufio.NewReader(stream)
+	// Create a buffer stream for non blocking read and write.
+	raw, err := reader.ReadString(DELIMITER)
+
+	// An EOF means the stream has been closed by the other side.
+	if errors.Is(err, io.EOF) {
 		stream.Close()
+		return
+	}
+	ms.checkError(err)
+	m, err := protocols.DeserializeMessage(raw)
+	ms.checkError(err)
+	ms.toEngine <- m
+	stream.Close()
+}
+
+func (ms *P2PMessageService) peerExchangeHandler(stream network.Stream) {
+	ms.receivePeerInfo(stream)
+	ms.sendPeerInfo(stream)
+	stream.Close()
+}
+
+// sendPeerInfo sends our peer info over the given stream
+func (ms *P2PMessageService) sendPeerInfo(stream network.Stream) {
+	raw, err := json.Marshal(basicPeerInfo{
+		Id:      ms.Id(),
+		Address: ms.me,
 	})
 
-	return ms
+	ms.checkError(err)
+	writer := bufio.NewWriter(stream)
+	// We don't care about the number of bytes written
+	_, err = writer.WriteString(string(raw) + string(DELIMITER))
+	ms.checkError(err)
+	writer.Flush()
+}
+
+// receivePeerInfo receives peer info from the given stream
+func (ms *P2PMessageService) receivePeerInfo(stream network.Stream) {
+	reader := bufio.NewReader(stream)
+	// Create a buffer stream for non blocking read and write.
+	raw, err := reader.ReadString(DELIMITER)
+
+	// An EOF means the stream has been closed by the other side.
+	if errors.Is(err, io.EOF) {
+		stream.Close()
+		return
+	}
+	ms.checkError(err)
+
+	var peerInfo *basicPeerInfo
+	err = json.Unmarshal([]byte(raw), &peerInfo)
+	ms.checkError(err)
+
+	prev, hasPeer := ms.peers.Load(peerInfo.Address.String())
+
+	if !hasPeer {
+		fmt.Printf("Setting info received from peer: %v\n", peerInfo)
+		ms.peers.Store(peerInfo.Address.String(), *peerInfo)
+		ms.newPeerInfo <- *peerInfo
+	}
+
+	// It's possible that the a message service is using a new peer ID for the same address.
+	// We allow overriding the existing peer info in this case.
+	if isNewInfo := prev.Id != peerInfo.Id || prev.Address != peerInfo.Address; hasPeer && isNewInfo {
```

Comment:
Not sure I fully understand this condition. `ms.peers.Load` lookup is done by address. So if the incoming peer information address does not match looked up peer address, I believe `hasPeer` will always be false?

Reply:
Removed in https://github.com/statechannels/go-nitro/pull/1207/commits/738c9ef14429f8ac423289a7c1f1ed7daaccb974